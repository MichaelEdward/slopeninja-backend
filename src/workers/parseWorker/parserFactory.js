import cheerio from 'cheerio';

export const removeBackSlashes = (text) => {
  const decoded = text ? text.toString().replace(/\\/g, '') : null;
  return decoded;
}

export const createHtmlParser = (key, parser, preParser = t => t) => async (html) => {
  const $ = cheerio.load(preParser(html));

  const keyData = await parser($);

  return {
    [key]: keyData
  };
}

export const createJSONParser = (key, parser) => async (data = "{}") => {
  let json = JSON.parse(data);

  let d;
  if (json) {
    d = json;
  }

  const keyData = await parser(d);

  return {
    [key]: keyData
  };
}
