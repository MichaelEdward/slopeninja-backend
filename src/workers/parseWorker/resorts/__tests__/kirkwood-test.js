import fs from 'fs';
import {
  parseKirkwoodSnow,
  parseKirkwoodLifts,
  parseKirkwoodLiftList,
  parseKirkwoodTrailList,
  parseKirkwoodTrails
} from '../kirkwood';
import { createHtmlParser } from '../../parserFactory';

test('fetches Kirkwood snow data correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/fixtures/kirkwood-weather.html.input`);

  const resortData = await createHtmlParser('snow', parseKirkwoodSnow)(htmlText);
  expect(resortData).toEqual({
    snow: {
      status: 'closed',
      weatherIcon: null,
      temprature: null,
      baseCondition: null,
      newSnow: 10,
      snowDepthBase: null,
      snowDepthSummit: 1,
    },
  });
})

test('fetches all null for nonexisting snow values', async () => {
  const resortData = await createHtmlParser('snow', parseKirkwoodSnow)('<html></html>');
  expect(resortData).toEqual({
    snow: {
      status: null,
      weatherIcon: null,
      temprature: null,
      baseCondition: null,
      newSnow: null,
      snowDepthBase: null,
      snowDepthSummit: null,
    }
  });
});

test('fetches Kirkwood lifts data correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/fixtures/kirkwood-lifts.html.input`);
  const resortData = await createHtmlParser('lifts', parseKirkwoodLifts)(htmlText);
  expect(resortData).toEqual({
    lifts: {
      total: 15,
      open: 0,
    }
  });
});

test('fetches all null for nonexisting lift values', async () => {
  const resortData = await createHtmlParser('lifts', parseKirkwoodLifts)('<html></html>');
  expect(resortData).toEqual({
    lifts: {
      total: null,
      open: null,
    }
  });
});

test('fetches Kirkwood trails data correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/fixtures/kirkwood-lifts.html.input`);
  const resortData = await createHtmlParser('trails', parseKirkwoodTrails)(htmlText);
  expect(resortData).toEqual({
    trails: {
      total: 86,
      open: 0,
    }
  });
});

test('fetches all null for nonexisting trails values', async () => {
  const resortData = await createHtmlParser('trails', parseKirkwoodTrails)('<html></html>');
  expect(resortData).toEqual({
    trails: {
      total: null,
      open: null,
    }
  });
});

test('fetches Kirkwood lift list correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/fixtures/kirkwood-lifts.html.input`);
  const resortData = await createHtmlParser('liftList', parseKirkwoodLiftList)(htmlText);
  expect(resortData).toMatchSnapshot();
});

test('fetches all null for nonexisting lift list values', async () => {
  const resortData = await createHtmlParser('liftList', parseKirkwoodLiftList)('<html></html>');
  expect(resortData).toMatchObject({ liftList: [] });
});

test('fetches Kirkwood trail list correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/fixtures/kirkwood-lifts.html.input`);
  const resortData = await createHtmlParser('trailList', parseKirkwoodTrailList)(htmlText);
  expect(resortData).toMatchSnapshot();
});

test('fetches all null for nonexisting lift list values', async () => {
  const resortData = await createHtmlParser('trailList', parseKirkwoodTrailList)('<html></html>');
  expect(resortData).toMatchObject({ trailList: [] });
});
