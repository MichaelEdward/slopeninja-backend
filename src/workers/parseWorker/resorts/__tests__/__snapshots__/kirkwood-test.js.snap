// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches Kirkwood lift list correctly 1`] = `
Object {
  "lifts": Array [
    Object {
      "category": null,
      "name": "Snowkirk",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "The Wall",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "The Reut",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Magic Carpet 1",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Magic Carpet 2",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Lookout Vista",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Covered Wagon 1",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Covered Wagon 2",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Caples Crest",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Iron Horse",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Sunrise",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Solitude",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Cornice Express",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "TC Express",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Red Cliffs Tow",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Bunny",
      "status": "closed",
    },
  ],
}
`;

exports[`fetches Kirkwood trail list correctly 1`] = `
Object {
  "trails": Array [
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Bud's Alley",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Easy Rider",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Elevator Shaft",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Happiness Is",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Herringbone Straight",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Lower Devil's Draw",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Button Bowl",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Cold Shoulder",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Covered Wagon",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Coyote",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Deadwood Spur",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Fawn Ridge",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Hully Gully",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Larry's Lip",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Sunnyside",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "The Wave",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Upper Devil's Draw",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "beginner",
      "name": "Bunny Gully",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "beginner",
      "name": "Ditch of Doom",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "beginner",
      "name": "Forbidden Forest",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "beginner",
      "name": "Funny Bunny",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "beginner",
      "name": "Gold Rush",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "beginner",
      "name": "Graduation",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "beginner",
      "name": "Hay Flat",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "beginner",
      "name": "School Run",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "beginner",
      "name": "Snowkirk",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "beginner",
      "name": "Squirrel Ridge",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "beginner",
      "name": "Trench of Terror",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Bark Shanty",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Buckboard",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Caples Crest Traverse",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Flying Carpet",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Free N Easy/Bandit Park",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Hole N Wall",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Hole N Wall Trees",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Home Run",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Jambi's Jaunt",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Jane's Jaunt",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Juniper",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Low Whiskey",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Lower Monte Wolfe",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Lower Zachary",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Mokelumne",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Outlaw Park",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Scott's Run",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Up N Over",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "intermediate",
      "name": "Whiskey Slide",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Conestoga",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Eagle Bowl",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Fireball",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Headwaters",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "High Eagle Bowl",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "High Whiskey",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "jeff's Line",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Jim's",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "KGF",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Lightning",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Look Out Janek",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Lost Cabin",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Lower Lookout Janek",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Lower Olympic",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Monte Wolfe",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Olympic",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Palisades Bowl",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Rabbit Runs",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Sentinel",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Sentinel Bowl",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Short Spoke",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Shotgun",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Snow Snake Gully",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Stump Run",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "The Drain",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Thunder Saddle",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Wagon Trail",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "advanced",
      "name": "Zachary",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "expert",
      "name": "All The Way",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "expert",
      "name": "Bogie's Slide",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "expert",
      "name": "Boylermaker",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "expert",
      "name": "Chamonix",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "expert",
      "name": "Cliff Chute",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "expert",
      "name": "Dick's Drop",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "expert",
      "name": "Hell's Delight",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "expert",
      "name": "Norms Nose",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "expert",
      "name": "Notch Chute",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "expert",
      "name": "One Man Chute",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "expert",
      "name": "Saddle Chute",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "expert",
      "name": "Shaffer's Chute",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "expert",
      "name": "Shot 5",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "expert",
      "name": "Sister Chute",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "expert",
      "name": "The Wall",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "expert",
      "name": "Two Man Chute",
      "status": "closed",
    },
    Object {
      "category": null,
      "level": "expert",
      "name": "Wagon Wheel Bowl",
      "status": "closed",
    },
  ],
}
`;
